content/blog/waddlesplash/2023-04-24_condition_variables.md

```diff
@@ -0,0 +1,416 @@
+void add() {
+	int32 localInteger = gSharedInteger;
+	localInteger = localInteger + 1;
+	gSharedInteger = localInteger;
+}
+```
+What happens if `add` is invoked for the first time simultaneously in two separate threads? Will `sharedInteger` become `2`, or (if the invocations really occur at exactly the same instant), will it become `1`, since the threads will have “raced” against each other?
+
+To solve this problem, ISAs (instruction set architectures) include _atomic operations_, so called because they act like one operation, not distinct ones. In our `add` function, we performed 3 distinct operations: we fetched the value, added `1` to it, and then set the value. With an atomic operation, these 3 operations become 1 single operation (for all intents and purposes of the programmer, anyway), and so what order threads run them in is irrelevant.
+
+Different compilers, operating systems, etc. expose such ISA-level atomic operations in different ways. On Haiku, we expose them through a set of functions beginning with `atomic_` which operate on `int32*`s (and there is also an equivalent set of functions for `int64*`s.) If we use one of them to make our earlier example thread-safe, it will look like this:
+```cpp
+void add() {
+	atomic_add(gSharedInteger, 1);
+}
+```
+(As a bonus, `atomic_add` also returns the value `gSharedInteger` had before `1` was added to it: so it is really a “fetch and add” operation, not just an “add” operation.)
+
+There are a variety of different atomic operations, and many ways to use them. But it is important to keep in mind that these are a ‘low-level’ primitive, and when using them, there are a lot of considerations that must be taken into account: when using multiple atomic operations in a row, interesting things can happen with memory ordering, for example.
+
+### Spinlocks
+
+Atomic operations mostly serve as a “building block” upon which more ergonomic synchronization primitives are built. Foremost among these, underlying all other thread synchronization primitives in the kernel, is the *spinlock*.
+
+```cpp
+spinlock gLock = B_SPINLOCK_INITIALIZER;
+int32 gSharedInteger = 0;
+
+void add() {
+	acquire_spinlock(&gLock);
+	gSharedInteger = gSharedInteger + 1;
+	release_spinlock(&gLock);
+}
+```
+
+As the name implies, spinlocks simply “spin” (run atomic operations in a loop) until they manage to acquire the lock. They can thus wind up wasting lots of CPU time if another core holds them for a long period of time. (Haiku’s implementation has mechanisms to detect this and trigger kernel panics when it happens.) They are mostly used in implementing other synchronization primitives, and in especially low-level parts of the kernel.
+
+Any thread which is acquiring or holding spinlocks is not allowed to be rescheduled, or even interrupted for any reason. Thus, spinlocks are always acquired only after the CPU’s “interrupts disabled” mode has been activated, and attempts to acquire them outside this mode will fault.
```

Comment:
I think understanding this requires a specific definition if "interrupted". I understand that it's about hardware interrupts, but it would be better to make this explicit?

Likewise, "fault" is not great here since it can mean a variety of things. In this case, talking about a kernel panic would avoid confusion (for example with CPU level faults like segmentation faults).